### ISSUE TITLE
Reranker speed optimizaiton

### ISSUE THREAD

**Issue by AmosChenZixuan** [AmosChenZixuan/BibiLab#698]
# Reranker speed optimization — spike results

**Status:** spike complete. Whoever picks this up creates an epic + sub-issues from the proposed actions.

## Bench result

Tooling: `backend/tests/bench_chat_prellm.py` (one-shot, delete after verify). Artifact: `~/.bibilab/bench/chat_prellm_20260905_034303.json`.

Setup: list `a52f764f…` (22 sources, densest in DB), N=150 synthetic queries (1–3 random chunks concatenated), 5 warmup, top_k=8.

| stage | mean ms | p50 | p95 | max |
|---|---:|---:|---:|---:|
| embed | 84 | 70 | 164 | 282 |
| hybrid | 413 | 386 | 816 | 1106 |
| **rerank** | **5293** | **5101** | **5848** | **6039** |
| total | 5706 | 5553 | 6363 | 6613 |

Candidate pool: mean 60.2 (cap hit), reranker output constant 8 → 7.5× over-kill.

## Findings

1. **Reranker's 512 ceiling is the silent binding constraint.** `pipeline/rerank.py:50` hardcodes `enable_truncation(max_length=512)`. Pair encoding (`[CLS] q [SEP] d [SEP]`, ~509 usable tokens). Truncation `longest_first`, no signal to caller. en chunks (≤400) fit when query ≤109 tokens; **zh chunks (≤1066) always truncated**; query >510 tokens → doc fully zeroed, cross-encoder scores empty content.

2. **Latency bottleneck is rerank (93%), not embed/hybrid.** embed is healthy for single query. hybrid's real cost is the **FTS raw-text backfill** at `pipeline/embed.py:466-468` (Chroma `get(ids=…)` to fetch raw texts after FTS5 hits, since FTS content is token-soup) — not HNSW (791 vectors, tiny).

3. **Retrieval is gateless hybrid + RRF + rerank.** Chroma vector (`max_distance=0.8` floor) ∥ FTS5 (CJK + pinyin) → RRF(k=60) → rerank(query, all candidates, top_k=8). No post-rerank gate; rerank is ordering, not authority.

4. **Chunks are derived from segments.** `transcript_segments` is canonical. Changing `pipeline/chunk.py` constants + re-running `embed_chunks()` per source = light backfill (re-embed only, no re-transcribe). 55 zh sources in current DB → backfill ~1.5 min estimated.

5. **Confirmed env facts:**
   - `chat_inference_pool.max_workers = 4` (`pipeline/chat_inference_pool.py:22`) — sufficient for single user
   - Embed model is **fp32** (`model_registry.py:116-134`, 470 MB) — unlike reranker which is int8 (266 MB)
   - Chroma collection = **791 chunks** total — far from any scalability concern
   - Chroma HNSW defaults (ef_search=10, M=16, ef_construction=100), not exposed via API

## Proposed next steps (5 items, medium-to-high ROI)

1. **zh chunk cap 800→400 + backfill** — change `_LANG_TARGET_TOKENS["zh"]` in `pipeline/chunk.py:16-19`; backfill script re-embeds 55 zh sources. Eliminates silent truncation quality risk; aligns zh with en.

2. **GPU rerank** — enable CUDA EP in `interpreting_providers()`. Single biggest perf lever (10–50×). Hardware-dependent; if no CUDA machine in scope, defer.

3. **Candidate-pool sweep bench** — parameterize `effective_top_k` (10/30/60) in `backend/tests/bench_chat_prellm.py`. Quantifies recall-vs-latency curve to inform production cap.

4. **Pre-fusion top_k 60→30 (vec/fts each)** — `pipeline/embed.py:505-506`. Rerank input drops from ~60 to ~40 candidates → rerank 5.3s → ~3.5s. Apply after the candidate-pool sweep (item 3 above) confirms the recall curve.

5. **FTS raw-text backfill async** — `pipeline/embed.py:466-468` to fire-and-forget; RRF falls back to FTS token-soup if backfill not yet done. Hybrid 413ms → ~200ms. Needs degradation-path testing.

## Explicitly excluded

- **Embed fp32→int8** — recall risk + < 1% chat perf gain → defer until measured.
- **Pool size 4→8** — only matters at 4+ concurrent chats; not observed.
- **Rerank-time doc truncation** — obviated by the zh cap change (item 1 above): zh cap ≤400, en already ≤400.

**Comment by AmosChenZixuan** [comment]
# Counter-measurement: finding 1 is a tokenizer-unit error, and it points the wrong way

**Environment caveat, read first.** These measurements are from the macOS dev machine, **not** the machine that produced the bench above. This home has 4 lists / 18 sources / 136 chunks / 2955 segments; `~/.bibilab/bench/` does not exist, list `a52f764f…` is not in the DB, and `backend/tests/bench_chat_prellm.py` is absent (deleted as the spike intended). So the latency table is **not reproducible here** and none of the timings below are re-measured — only the token-length claims are. Whoever resumes on the bench machine should re-run the commands at the bottom before acting.

## The unit mismatch

`pipeline/chunk.py` sizes chunks with `_shared.count_tokens` → **cl100k_base**. The reranker consumes **XLM-R sentencepiece** (Unigram, 250002 vocab). These are different units, and the ratio is language-dependent in the opposite direction to what finding 1 assumes.

Measured across all 2955 segments, grouped by source (ratio = XLM-R tokens ÷ cl100k tokens):

| text | n sources | ratio p50 | ratio range | 800 cl100k → XLM-R |
|---|---:|---:|---|---:|
| Chinese | 16 | **0.576** | 0.569 – 0.604 | **461** |
| English | 2 | **1.082** | 1.057 – 1.106 | **865** |

XLM-R is ~1.7× *more* token-efficient on Chinese than cl100k, and ~10% *less* efficient on English.

**Consequence: `_LANG_TARGET_TOKENS["zh"] = 800` produces ~461 XLM-R tokens — comfortably under the ~509 usable ceiling. Chinese chunks do not truncate.** Finding 1's "zh chunks (≤1066) always truncated" measures the document in a tokenizer the reranker does not use.

Per-chunk distribution over the 136 chunks in the Chroma collection:

| | cl100k p50 | XLM-R p50 | XLM-R p95 | XLM-R max | % over 509 XLM-R |
|---|---:|---:|---:|---:|---:|
| zh (n=130) | 484 | 288 | 485 | 560 | **1.5%** |
| en (n=6) | 468 | 522 | 544 | 807 | **50%** |

## What actually truncates, and why

Every over-ceiling chunk traces to two English-audio sources labeled `language='zh'`:

```
a48c9458  language=zh  cjk=0.01  ratio=1.11  "Reading China's Propaganda – What They Really…"
dab06ae3  language=zh  cjk=0.03  ratio=1.06  "Freak athletes 🦘@Dakota Rivers…"
```

Chain: `~/.bibilab/config.json` sets `transcription.language: "zh"` → `worker.py:547-551` honors it for sensevoice and assigns `effective_language = detected_language` → `worker.py:595` passes it to `chunk_by_sections` → `chunk.py:70` selects the Chinese 800-token budget **for English text** → ~865 XLM-R tokens → real truncation at 512.

So the defect is not the reranker's ceiling and not Chinese. It is that chunk sizing depends on a **two-hop proxy** — ASR language label → language → cl100k budget — standing in for the one question that matters: *how many reranker tokens will this be?* The label is a user-forced ASR setting, not a property of the text, so the proxy breaks whenever the two disagree.

Affected population here: **5 of 136 chunks (3.7%)**, all from that config mismatch.

## Available fix, and why it is cheap

The embedding model's tokenizer is byte-identical in behaviour to the reranker's, and it is already present at ingest time (the reranker is not fetched until first chat):

```
identical token counts: 118 / 118      max abs diff: 0
```

Both are XLM-R Unigram / 250002 vocab. So `chunk.py` can size in the exact unit that binds, using a tokenizer guaranteed available when chunking runs. `_LANG_TARGET_TOKENS` and the `language` parameter both become unnecessary — one target (~480 XLM-R tokens) serves every language and cannot be defeated by a wrong label. Affected sources need a re-chunk + re-embed backfill (segments are canonical, no re-transcription).

## Correction to proposed step 1

> **zh chunk cap 800→400 + backfill** … Eliminates silent truncation quality risk; aligns zh with en.

Two problems. It does not eliminate a risk that the measurements say is not present for Chinese. And "aligns zh with en" inverts `chunk.py:11-14`'s own rationale — Chinese is given a *larger* cl100k target precisely to hold a *comparable* semantic span (~1 tok/char vs ~1 tok/4 chars). Capping zh at 400 cl100k makes a Chinese chunk carry roughly half an English chunk's content, and drives real XLM-R length down to ~230 against a 509 ceiling — paying semantic span for headroom that is already there.

## Correction to the latency arithmetic

Two errors worth recording before anyone sizes work off them:

- **Longer sequences cost ~2.3×, not ~4×.** At 512–1066 tokens in a 12L/768H encoder, attention is only ~10% of FLOPs; projections and FFN scale linearly with length, attention quadratically. Full-doc reranking at ~1066 tokens is ≈2.3× a 512-token pair, not 4×.
- **Parameter count is a poor cost proxy across XLM-R-family rerankers.** bge-reranker-base's 278M is ~192M embedding lookup + ~86M encoder; `gte-multilingual-reranker-base`'s 306M is the same 12L/768H encoder with the same 250k vocab. Per-token compute is essentially equal, so a swap to an 8192-ctx model would be compute-neutral *per token* — only the longer sequences would cost.

Also unexamined and free: `rerank.py:56` pads every pair to the batch maximum. Real XLM-R lengths here are ~288 p50, so a handful of long chunks drag the whole batch to the ceiling. Length-bucketed batching is a pure win, no model change and no quality risk.

## Adjacent finding, larger than this issue

The embedding tokenizer ships `truncation.max_length: 128` — the trained sequence length of `paraphrase-multilingual-MiniLM-L12-v2`. `pipeline/embed.py:186` overrides it to **512**. Every chunk in the DB is therefore embedded across positions 128–512 that the sentence-transformers fine-tune never calibrated. If that degrades embedding quality it affects the vector arm of *every* query and outweighs reranker truncation entirely — and it changes what chunk size should be in the first place. Not investigated here; worth its own spike before chunk-size decisions are locked.

## Reproduce / refute on the bench machine

```python
# ratio of XLM-R (what the reranker consumes) to cl100k (what the chunker counts)
import os, glob, sqlite3, statistics, tiktoken
from tokenizers import Tokenizer
tok = Tokenizer.from_file(glob.glob(os.path.expanduser(
    '~/.bibilab/models/reranker/**/tokenizer.json'), recursive=True)[0])
enc = tiktoken.get_encoding('cl100k_base')
c = sqlite3.connect(os.path.expanduser('~/.bibilab/bibilab.db')); c.row_factory = sqlite3.Row
for s in c.execute('select id, language, substr(title,1,30) t from sources'):
    segs = [r[0] for r in c.execute(
        'select text from transcript_segments where source_id=?', (s['id'],))]
    if not segs: continue
    txt = ' '.join(segs)
    print(s['t'], s['language'], round(len(tok.encode(txt).ids) / len(enc.encode(txt)), 3))

# embedder vs reranker tokenizer equivalence
et = Tokenizer.from_file(os.path.expanduser('~/.bibilab/models/embedding/onnx/tokenizer.json'))
et.no_truncation()   # tokenizer.json pins max_length=128; embed.py overrides to 512
```

## Open questions for whoever resumes

1. Does the ratio hold on the 22-source list (expect zh ≈ 0.58, en ≈ 1.08)? If yes, finding 1 and proposed step 1 both need withdrawing.
2. What is the over-509 XLM-R population on that DB — is it also only mislabeled-language sources?
3. Restore a permanent bench before any latency sub-issue; the current figures cannot be reproduced or regression-tested.
4. Spike the embedder 128→512 override before locking any chunk-size target.

**Comment by AmosChenZixuan** [comment]
Closing. Re-measured on the machine that produced the original bench. Finding 1 is wrong, proposed
step 1 would have made things worse, and the bench measured a query shape that never occurs in
production. The correctness work has moved to epic #713. The perf evidence is recorded below, since
closing this issue is otherwise where it would be lost.

## Finding 1 — the unit error is real, but neither this issue nor the counter-comment got the
## population right

The counter-comment's ratios reproduce here exactly: XLM-R ÷ cl100k is **zh 0.578** (range
.530–.634, n=53) and **en 1.081** (n=2), and the embedding tokenizer is behaviourally identical to
the reranker's (token counts match on 55/55 sources). So `zh = 800` cl100k is ~462 XLM-R, and
**"zh chunks (≤1066) always truncated" is false.**

But that comment's own conclusion — *"1.5% of zh chunks over the ceiling, every one traceable to two
mislabeled sources"* — does not hold here either. On this corpus **8.5% of Chinese chunks exceed 509
XLM-R tokens, and 50 of the 52 over-ceiling chunks are genuine Chinese.** The difference is corpus,
not method: that box's chunks were cl100k p50 484 against 820 here.

The number that actually matters was measured by neither: over 65 real queries × 120 real chunks,
**20.4% of rerank pairs exceed the 512 window**, mean 23 tokens dropped.

And there is a second truncation site nobody looked at. `embed.py:186` also caps at 512, so **44 of
593 chunks (7.4%) are embedded with their tails missing**, worst losing 320 tokens. That loss is
baked into the stored vector — those passages are unreachable by any query, which is strictly worse
than the per-query rerank truncation this issue was about.

## Finding 1's mechanism is also wrong

> query >510 tokens → doc fully zeroed, cross-encoder scores empty content

`longest_first` trims **both** sides. Measured with a two-chunk query: `query=254 / doc=254`. The
document is halved, not zeroed.

## Proposed step 1 (zh 800→400 + backfill) — withdrawn

It would have paid semantic span for headroom that largely existed, and it cannot work in principle:
the worst source in the corpus is **mixed zh/en** (`dbecd18a`, CJK ratio 0.670, **60%** of its chunks
over the ceiling) — worse than either pure-English source. Truncation varies continuously with CJK
ratio, so no target keyed on a language label can represent it. The label is a user-forced ASR config
value anyway (#703).

Replaced by #716 (size chunks in the reranker's own tokenizer, delete the language keying, enforce a
real ceiling) and #714 (explicit query + document budget).

## The bench measured a pair shape that does not occur

Queries were "1–3 random chunks concatenated". Per the measurement above, such a query is truncated
to 254 tokens and takes half the document with it. Real `find_passages` queries — 116 of them,
extracted from 83 recorded turns in `~/.bibilab/debug/` — are XLM-R **p50 14, p95 26, max 51**.

Total pair length lands near 512 either way, so the *timings* are roughly right. Nothing about the
bench's content reflects production, and the harness is gone, so none of it is regression-testable.

## Perf evidence, for whoever picks this up (deferred, no open tracker)

Reproduced and extended on this box (i9-14900F 32T, RTX 5070 Ti, ORT 1.29):

| Config | p50 |
|---|---|
| int8 CPU, 60×512 (today) | 4419–4950 ms |
| Thread sweep 8/12/16/24/32 | 4.4–4.9 s — **flat, no tuning headroom** |
| fp32 CPU, 60×512 | 8153 ms (1.85× slower than int8) |
| **fp32 CUDA, 60×512, real queries** | **229 ms p50 / 258 p95** (load 1.8 s, +3.7 GB VRAM) |
| **int8 CUDA** | **8346 ms — slower than on CPU** |
| CPU pool sweep (doc 448) | 60 → 4419 · 40 → 3027 · 24 → 1889 · 16 → 1269 ms |

Three things that change the proposed steps:

1. **Step 2 as written is a regression.** `DynamicQuantizeMatMul` has no efficient CUDA kernel, so
   flipping the execution provider on the shipped int8 spec makes it 1.7× *slower*. `interpreting_providers()`
   already returns CUDA first when available — the real blockers are the CPU-only `onnxruntime` pin in
   `pyproject.toml` and the fact that only the int8 spec ships. The GPU path needs the fp32 spec back.
2. **Rerank is purely token-bound** at ~6.3–7 tokens/ms; op profile is 51% int8 linear layers, ~22%
   attention. There is no configuration win — only fewer or shorter pairs.
3. **Length-bucketed batching is not a lever.** Padding waste over 40 real candidate pools is p50
   **16.9%**, max 23.5%.

Steps 3 and 4 (candidate-pool sweep, pre-fusion top_k 60→30) remain valid and unmeasured for recall.

Also worth correcting while it is in view: `model_registry.py:135` claims the int8 reranker keeps
"top-8 91% vs fp32". Measured over 20 real queries at pool 60: **top-8 overlap 6.70/8 = 83.75%**, and
top-3 only 2.20/3 = 73% — the loss concentrates in the highest ranks.

## Where it went

- Correctness → epic **#713** (#714 pair budget, #715 embedding model swap, #716 chunk unit +
  enforced ceiling, #717 truncation telemetry, #718 backfill), plus **#703** for the language label.
- Perf → deferred by decision, with the evidence above.

Full measurement record and reproduction recipes: `docs/handoff/reranker-chunk-unit-session.md`.

**Comment by AmosChenZixuan** [comment]
Superseded by #713 for the correctness work; perf deferred with its evidence recorded in the correction above.
